Implement Phase2 HLT Taus and L1HPSTaus

sandeepbhowmik1 opened the pull request:
#### PR description:

<!-- Please replace this text with a description of the feature proposed or problem addressed, specifying:
  - what changes are expected in the output if any, 
  - what other PRs or externals it depends upon if any,
  - link to any additional material useful to provide a documentation for this PR (slides, JIRA tickets, related pull requestes, hypernews, TWiki or Indico pages)  -->

#### PR validation:

<!-- Please replace this text with a description of which tests have been performed to verify the correctness of the PR, including the eventual addition of new code for testing like unit tests, test configurations, additions or updates to the runTheMatrix test workflows -->

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

<!-- Please replace this text with any link to  -->

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

fwyzard commented:
@sandeepbhowmik1 it looks like the changes to RecoTauTag/RecoTau overwrite and/or revert some of the changes made by #32831 (11.1.x) / #31744 (11.2.x and later).
Could you update the HLT part to make use of what is already in 11.1.8 - or add new code if needed, but without reverting what is already in the release ?

fwyzard commented on DataFormats/Phase2HLTPFTaus/interface/PFTauPair.h:
```diff
@@ -0,0 +1,39 @@
+    ~PFTauPair();
+
+    const reco::PFTauRef& leadPFTau() const;
+    double leadPFTau_sumChargedIso() const;
```
Please use camelCase (preferred), i.e. `leadPFTauSumChargedIso()` or snake_case (e.g. `lead_PFTau_sum_charged_iso()`, but not both at the same time.

fwyzard commented on DataFormats/Phase2HLTPFTaus/interface/PFTauPairFwd.h:
```diff
@@ -0,0 +1,13 @@
+
+namespace reco
+{
+  typedef std::vector<PFTauPair> PFTauPairCollection;
```
suggestion:
```suggestion
  using PFTauPairCollection = std::vector<PFTauPair>;
```

fwyzard commented on DataFormats/Phase2HLTPFTaus/interface/PFTauPairFwd.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef DataFormats_Phase2HLTPFTaus_PFTauPairFwd_h
```
Alternatively, drop the `#include "DataFormats/Phase2HLTPFTaus/interface/PFTauPair.h"` statement and use an actual forward declaration for `PFTauPair`, _e.g._
```c++
namespace reco {
    class PFTauPair;
    using PFTauPairCollection = std::vector<PFTauPair>;
}
```
though I'm not sure if it will work, or if it is actually useful.

fwyzard commented on DataFormats/Phase2HLTPFTaus/BuildFile.xml:
```diff
@@ -0,0 +1,8 @@
+<use   name="DataFormats/Candidate"/>
```
Do we really need a new package for a single class ?
Can it go in an existing one, for example `DataFormats/TauReco` or `DataFormats/HLTReco` ?

fwyzard commented on DataFormats/Phase2HLTPFTaus/src/PFTauPair.cc:
```diff
@@ -0,0 +1,143 @@
+  assert(reco::deltaR(leadPFTau_->p4(), subleadPFTau_->p4()) > 1.e-3);
+}
+
+PFTauPair::~PFTauPair()
+{}
```
Remove it from here and declare it as `= default;` in the header file.

fwyzard commented on DataFormats/Phase2HLTPFTaus/src/PFTauPair.cc:
```diff
@@ -0,0 +1,143 @@
+
+namespace reco {
+
+PFTauPair::PFTauPair()
```
Remove it from here and define it as = default; in the header file.

fwyzard commented on DataFormats/Phase2HLTPFTaus/src/PFTauPair.cc:
```diff
@@ -0,0 +1,143 @@
+PFTauPair::PFTauPair(const reco::PFTauRef& leadPFTau, double leadPFTau_sumChargedIso, const reco::PFTauRef& subleadPFTau, double subleadPFTau_sumChargedIso, double dz)
+  : LeafCandidate(leadPFTau->charge() + subleadPFTau->charge(), 
+                  leadPFTau->p4() + subleadPFTau->p4(), 
+                  reco::LeafCandidate::Point(0.5*(leadPFTau->vertex().x() + subleadPFTau->vertex().x()), 
+                                             0.5*(leadPFTau->vertex().y() + subleadPFTau->vertex().y()), 
+                                             0.5*(leadPFTau->vertex().z() + subleadPFTau->vertex().z())))
```
question: does it work to write
```suggestion
                  (leadPFTau->vertex() + subleadPFTau->vertex()) / 2.)
```
instead ?

fwyzard commented on HLTrigger/Phase2HLTPFTaus/plugins/PFRecoTauChargedHadronQualityPluginHGCalWorkaround.h:
```diff
@@ -0,0 +1,16 @@
+{
+ public:
+  explicit PFRecoTauChargedHadronQualityPluginHGCalWorkaround(const edm::ParameterSet& cfg);
+  ~PFRecoTauChargedHadronQualityPluginHGCalWorkaround() override {}
```
```suggestion
  ~PFRecoTauChargedHadronQualityPluginHGCalWorkaround() override = default;
```

fwyzard commented on HLTrigger/Phase2HLTPFTaus/plugins/PFRecoTauChargedHadronQualityPluginHGCalWorkaround.cc:
```diff
@@ -0,0 +1,27 @@
+#include "HLTrigger/Phase2HLTPFTaus/plugins/PFRecoTauChargedHadronQualityPluginHGCalWorkaround.h"
```
Maybe this plugin should go into `RecoTauTag/RecoTau/` ?

fwyzard commented on HLTrigger/Phase2HLTPFTaus/plugins/PFTauPairProducer.cc:
```diff
@@ -0,0 +1,108 @@
+#include "HLTrigger/Phase2HLTPFTaus/plugins/PFTauPairProducer.h"
```
Maybe this plugin should go under `RecoTauTag/HLTProducers/` ?

fwyzard commented on HLTrigger/Phase2HLTPFTaus/plugins/PFTauPairProducer.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef HLTrigger_Phase2HLTPFTaus_PFTauPairProducer_h
```
Can you merge the class declaration into the .cc file ?

fwyzard commented on HLTrigger/Phase2HLTPFTaus/plugins/PFTauPairProducer.cc:
```diff
@@ -0,0 +1,108 @@
+
+void PFTauPairProducer::produce(edm::Event& evt, const edm::EventSetup& es)
+{
+  std::unique_ptr<reco::PFTauPairCollection> pfTauPairs(new reco::PFTauPairCollection());
```
```suggestion
  auto pfTauPairs = std::make_unique<reco::PFTauPairCollection>();
  ```

fwyzard commented on HLTrigger/Phase2HLTPFTaus/plugins/RecoTauCleanerPluginHGCalWorkaround.h:
```diff
@@ -0,0 +1,17 @@
+{
+ public:
+  explicit RecoTauCleanerPluginHGCalWorkaround(const edm::ParameterSet& cfg, edm::ConsumesCollector&& cc);
+  ~RecoTauCleanerPluginHGCalWorkaround() override {}
```
```suggestion
  ~RecoTauCleanerPluginHGCalWorkaround() override = default;
```

sandeepbhowmik1 commented:
Dear Andrea,

I have modified all the code considering your all comments.

There were merge conflict with some files in 
RecoTauTag/RecoTau

RecoTauTag/RecoTau/interface/DeepTauBase.h
RecoTauTag/RecoTau/plugins/DeepTauId.cc
RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIsolation.cc
RecoTauTag/RecoTau/src/DeepTauBase.cc

I have kept these from default i.e. which are in CMSSW_11_1_8

and I have renamed the new files updated by Christian 

RecoTauTag/RecoTau/interface/DeepTauBaseForHLT.h
RecoTauTag/RecoTau/plugins/DeepTauIdForHLT.cc
RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationByIsolationForHLT.cc
RecoTauTag/RecoTau/src/DeepTauBaseForHLT.cc

fwyzard commented:
Hi Sandeep,
yes, those are the conflicts I was talking about earlier.

I don't think having two different versions of the same modules of a good
idea.

Can you work with Christian to update the HLT set up to use the version in
the release ?  Or, if anything is missing in the code, to update it
accordingly ?

Thank you,
.Andrea

sandeepbhowmik1 commented:
ok I will talk to Christian about it.

sandeepbhowmik1 commented:
The present code can reproduce the HLT TDR results but still there are 2 technical issues to solve.
I have uploaded this to share and discuss with the experts.
So if anyone wants to use this code now please let me know.
Thanks

sandeepbhowmik1 commented:
So those 2 technical issues are also solved now
